### PR TITLE
[Feature]: Lora support

### PR DESCRIPTION
**Pull request by princepride**
## Summary
https://github.com/vllm-project/vime/issues/206
This PR adds end-to-end LoRA support for Megatron actor training with colocated vLLM rollout. When `--lora-rank > 0`, the training actor applies LoRA adapters via Megatron-Bridge, exports PEFT-format adapters after each weight update, and hot-loads them into vLLM through the runtime LoRA API. Rollout requests then use the adapter name instead of the base model path.

This enables parameter-efficient RL fine-tuning without full-weight sync between trainer and inference engines on every step.

## Motivation

Full-weight synchronization between Megatron and colocated vLLM engines is expensive for large models. LoRA reduces the sync payload to small adapter weights and lets vLLM load adapters at runtime, which is a better fit for iterative RL training.

### Training (Megatron)

- **New CLI args**: `--lora-rank`, `--lora-alpha`, `--lora-dropout`, `--lora-type` (`lora` / `canonical_lora`), `--target-modules`, `--exclude-modules`, `--lora-adapter-name`
- **`lora_utils.py`**: target module parsing/conversion (HF ↔ Megatron), PEFT config generation, adapter export (`adapter_model.bin` + `adapter_config.json`)
- **`model_provider.py`**: apply LoRA directly after Bridge provider returns the actor model (instead of relying on pre-wrap hooks)
- **`actor.py`**: wake up offloaded model before LoRA export in `offload_train + colocate` mode
- **`arguments.py`**: auto-enable `vllm_enable_lora`, validate LoRA requires `--megatron-to-hf-mode bridge` and `--colocate`

### Weight Sync (Trainer → vLLM)

- **`update_weight_from_tensor.py`**: when LoRA is enabled, skip full-weight IPC sync; export adapter and call `load_lora_adapter` on all rollout engines
- **`vllm_engine.py`**:
  - set `VLLM_ALLOW_RUNTIME_LORA_UPDATING=1` when LoRA is enabled
  - add `load_lora_adapter()` with unload-then-load flow
  - handle vLLM's non-JSON / empty 200 responses on LoRA endpoints

### Rollout

- **`vllm_rollout.py` / `vllm_streaming_rollout.py`**: use adapter name (default `vime_lora`) as the `"model"` field in rollout requests when LoRA is enabled

### Cleanup & Reliability

- **`train.py`**: wrap main loop in `try/finally` to dispose actor/rollout resources
- **`actor_group.py` / `actor.py`**: add `dispose()` for explicit wandb cleanup in Ray actors
- **`logging_utils.py`**: call `wandb.teardown()` after `wandb.finish()` to avoid `BrokenPipeError` noise on Ray actor exit

### Tests

```bash
python3 train.py \
  --hf-checkpoint /root/models/Qwen3-8B \
  --ref-load /root/models/Qwen3-8B_torch_dist \
  --load /root/models/Qwen3-8B \
  --finetune \
  --no-load-optim \
  --no-load-rng \
  --save /root/vime_lora_runs/qwen3-8b-dapo-lora-smoke \
  --save-interval 1 \
  --megatron-to-hf-mode bridge \
  --actor-num-nodes 1 \
  --actor-num-gpus-per-node 1 \
  --colocate \
  --calculate-per-token-loss \
  --prompt-data /root/datasets/dapo-math-17k/dapo-math-17k.jsonl \
  --input-key prompt \
  --label-key label \
  --apply-chat-template \
  --rollout-shuffle \
  --rm-type deepscaler \
  --num-rollout 1 \
  --rollout-batch-size 1 \
  --n-samples-per-prompt 1 \
  --rollout-max-response-len 1024 \
  --rollout-temperature 0.8 \
  --global-batch-size 1 \
  --micro-batch-size 1 \
  --advantage-estimator grpo \
  --kl-loss-coef 0.00 \
  --kl-loss-type k1 \
  --kl-coef 0.00 \
  --entropy-coef 0.00 \
  --eps-clip 4e-4 \
  --optimizer adam \
  --lr 1e-6 \
  --lr-decay-style constant \
  --weight-decay 0.1 \
  --adam-beta1 0.9 \
  --adam-beta2 0.98 \
  --tensor-model-parallel-size 1 \
  --pipeline-model-parallel-size 1 \
  --context-parallel-size 1 \
  --expert-model-parallel-size 1 \
  --expert-tensor-parallel-size 1 \
  --use-dynamic-batch-size \
  --max-tokens-per-gpu 4096 \
  --rollout-num-gpus-per-engine 1 \
  --rollout-num-gpus 1 \
  --vllm-gpu-memory-utilization 0.45 \
  --vllm-max-cudagraph-capture-size 32 \
  --lora-rank 8 \
  --lora-alpha 16 \
  --lora-dropout 0.0 \
  --lora-type lora \
  --target-modules all-linear \
  --only-train-params-name-list lora_A lora_B linear_in linear_out \
  --lora-adapter-name vime_lora \
  --attention-dropout 0.0 \
  --hidden-dropout 0.0 \
  --accumulate-allreduce-grads-in-fp32 \
  --attention-softmax-in-fp32 \
  --attention-backend flash \
  --use-wandb \
  --wandb-host https://api.wandb.ai \
  --wandb-project vime-lora-smoke \
  --wandb-group qwen3-8b-dapo-single-h100 \
  --wandb-key "${WANDB_API_KEY}" \
  --disable-wandb-random-suffix
```

### CODE REVIEW

**Comment by princepride** [comment]
Superseded by #336, re-opened from the fork branch `princepride:lora-support` (same commits). Closing this one.
